pyfr/nputil.py

```diff
@@ -126,3 +127,187 @@ def npdtype_to_ctypestype(dtype):
+        #rhs     
+        d[i] -= ai*dj
+        d[i] /= bi1
+
+    #backward substitution
+    dd = d[::-1]
+    for i, (ci, dj) in enumerate(zip(c[::-1], d[:0:-1]), start=1):
+        dd[i] -= ci*dj
+
+
+#tridiagonal solver for cyclic/periodic systems 
+#Using Sherman-Morrison formula and Thompson algorithm
+# A = A + u ⊗ v
+# u, v are vectors
+def trislvcyc(a, b, c, d, alpha, beta):
+    u = np.zeros(d.shape[0], dtype=d.dtype)
+    #Solve the two systems, Ax = d, Bz = u in place, i.e., d->x, u->z
+    gam = -b[0] 
+    b[0] *= 2
+    b[-1] -= alpha*beta/gam
+    u[0] = gam
+    u[-1] = alpha
+    #solve the two systems in one loop
+    #forward elimination
+    d[0] /= b[0]
+    u[0] /= b[0]
+    for i, (ai, bi, bj, ci, dj, uj) in enumerate(
+                zip(a, b[1: ], b[: -1], c, d[: -1], u[: -1]), start=1):
+        #lhs coeffs
+        c[i-1] = cdivb = ci/bj
+        b[i]  = bi1 = bi - ai*cdivb  
+        # first rhs
+        d[i] -= ai*dj
+        d[i] /= bi1
+        # second rhs
+        u[i] -= ai*uj
+        u[i] /= bi1
+
+    #backward substitution
+    dd = d[::-1]
+    uu = u[::-1]
+    for i, (ci, dj, uj) in enumerate(
+            zip(c[::-1], d[:0:-1], u[:0:-1]), start=1):
+        dd[i] -= ci*dj
+        uu[i] -= ci*uj
+
+    #compute the final solution
+    fact = (d[0] + alpha*d[-1]/gam)/(1.0 + u[0] + alpha*u[-1]/gam)
+    d -= fact*u[:, None]
+
+    
+class CurveFit:
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def intrp(self, *args):
+        pass
+
+class LinearFit(CurveFit):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._interp = np.vectorize(np.interp, signature='(r),(n),(m)->(r)')
+
+    def intrp(self, x_intrp, xs, ys):
+        return self._interp(x_intrp, xs, ys.T).T
+
+
+class CubicSplineFit(CurveFit):
+    def __init__(self, xs=None, ys=None, bctype='not-a-knot'):
+        super().__init__(xs=None, ys=None, bctype='not-a-knot')
+
+        self._bctype = bctype
+        if np.any(xs):
+            self._construct(xs, ys, self._bctype)
+                
+    def update(self, xs=None, ys=None, bctype='not-a-knot'):
+        compute = np.any(xs) or (not (bctype == self._bctype))
+        self._bctype = bctype   
+        if compute:
+            if not np.any(xs):
+                xs, ys, _ = self.spl
+            self._construct(xs, ys, self._bctype)
+
+    def _construct(self, xs, ys,  bctype='not-a-knot'):
+        self._xs = xs
+        self._ys = ys
+        a, b, c, d = self._spl_matrix_coeffs()
+        if bctype == 'periodic':
+            self._periodic(a, b, c, d)
+        elif bctype == 'not-a-knot':
+            self._notaknot(a, b, c, d)
+        self._zs = d
+
+    def intrp(self, x_intrp, xs=None, ys=None):
+        if np.any(xs):
+            self.update(xs, ys, self._bctype)
+        return self._spleval(x_intrp)
+
+    @property
+    def spl(self):
+        return tuple((self._xs, self._ys, self._zs))
```

Comment:
This line is pointless; `(self._xs, self._ys, self._zs)` is a tuple so passing it to `tuple` does nothing.  Probably better just to have `self.spl = (x, y, z)` or something along those lines.